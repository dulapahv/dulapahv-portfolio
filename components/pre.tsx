'use client';

import { DetailedHTMLProps, HTMLAttributes, useRef, useState } from 'react';

import { Check, Copy } from 'lucide-react';

import { cn } from '@/lib/utils';

export default function Pre({
  children,
  ...props
}: DetailedHTMLProps<HTMLAttributes<HTMLPreElement>, HTMLPreElement>) {
  const [isCopied, setIsCopied] = useState(false);
  const preRef = useRef<HTMLPreElement>(null);

  const handleClickCopy = async () => {
    const code = preRef.current?.textContent;

    if (code) {
      await navigator.clipboard.writeText(code);
      setIsCopied(true);

      setTimeout(() => {
        setIsCopied(false);
      }, 1000);
    }
  };

  return (
    <>
      <button
        disabled={isCopied}
        onClick={handleClickCopy}
        aria-label="Copy code"
        title="Copy code"
        className={cn(
          `text-foreground-muted absolute top-[5px] right-2 cursor-pointer rounded-md p-2
          !transition-all`,
          'hover:text-foreground/80',
          'active:scale-90',
        )}
      >
        {isCopied ? <Check className="size-4" /> : <Copy className="size-4" />}
      </button>
      <pre ref={preRef} {...props}>
        {children}
      </pre>
    </>
  );
}
