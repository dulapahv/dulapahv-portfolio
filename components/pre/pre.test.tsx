import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import { Pre } from './pre';

describe('Pre', () => {
  // Mock clipboard API
  const mockWriteText = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
    mockWriteText.mockResolvedValue(undefined);

    Object.assign(navigator, {
      clipboard: {
        writeText: mockWriteText
      }
    });
  });

  describe('Component rendering', () => {
    it('should render pre element', () => {
      const { container } = render(
        <Pre>
          <code>const test = true;</code>
        </Pre>
      );

      const preElement = container.querySelector('pre');
      expect(preElement).toBeInTheDocument();
    });

    it('should render children inside pre element', () => {
      const { container } = render(
        <Pre>
          <code>const test = true;</code>
        </Pre>
      );

      const codeElement = container.querySelector('code');
      expect(codeElement).toBeInTheDocument();
      expect(codeElement?.textContent).toBe('const test = true;');
    });

    it('should render copy button', () => {
      render(
        <Pre>
          <code>const test = true;</code>
        </Pre>
      );

      const copyButton = screen.getByRole('button', { name: /copy code/i });
      expect(copyButton).toBeInTheDocument();
    });

    it('should render CopyIcon by default', () => {
      render(
        <Pre>
          <code>const test = true;</code>
        </Pre>
      );

      const copyButton = screen.getByRole('button', { name: /copy code/i });
      const svg = copyButton.querySelector('svg');

      expect(svg).toBeInTheDocument();
    });

    it('should pass props to pre element', () => {
      const { container } = render(
        <Pre className="custom-class" data-testid="custom-pre">
          <code>test</code>
        </Pre>
      );

      const preElement = container.querySelector('pre');
      expect(preElement).toHaveClass('custom-class');
      expect(preElement).toHaveAttribute('data-testid', 'custom-pre');
    });
  });

  describe('Copy functionality', () => {
    it('should copy code to clipboard when button is clicked', async () => {
      render(
        <Pre>
          <code>const test = true;</code>
        </Pre>
      );

      const copyButton = screen.getByRole('button', { name: /copy code/i });
      fireEvent.click(copyButton);

      await waitFor(() => {
        expect(mockWriteText).toHaveBeenCalledWith('const test = true;');
      });
    });

    it('should copy all text content from pre element', async () => {
      render(
        <Pre>
          <code>
            {`function hello() {
  console.log("Hello World");
}`}
          </code>
        </Pre>
      );

      const copyButton = screen.getByRole('button', { name: /copy code/i });
      fireEvent.click(copyButton);

      await waitFor(() => {
        expect(mockWriteText).toHaveBeenCalledWith(expect.stringContaining('function hello()'));
        expect(mockWriteText).toHaveBeenCalledWith(
          expect.stringContaining('console.log("Hello World")')
        );
      });
    });

    it('should show CheckIcon after successful copy', async () => {
      render(
        <Pre>
          <code>const test = true;</code>
        </Pre>
      );

      const copyButton = screen.getByRole('button', { name: /copy code/i });
      fireEvent.click(copyButton);

      // Button should now show check icon
      await waitFor(() => {
        const svg = copyButton.querySelector('svg');
        expect(svg).toBeInTheDocument();
      });
    });

    it('should disable button while copied state is active', async () => {
      render(
        <Pre>
          <code>const test = true;</code>
        </Pre>
      );

      const copyButton = screen.getByRole('button', { name: /copy code/i });
      fireEvent.click(copyButton);

      await waitFor(() => {
        expect(copyButton).toBeDisabled();
      });
    });

    it('should not copy if pre element has no text content', async () => {
      render(<Pre>{null}</Pre>);

      const copyButton = screen.getByRole('button', { name: /copy code/i });
      fireEvent.click(copyButton);

      expect(mockWriteText).not.toHaveBeenCalled();
    });
  });

  describe('Button attributes', () => {
    it('should have correct aria-label', () => {
      render(
        <Pre>
          <code>test</code>
        </Pre>
      );

      const copyButton = screen.getByRole('button', { name: /copy code/i });
      expect(copyButton).toHaveAttribute('aria-label', 'Copy code');
    });

    it('should have title attribute', () => {
      render(
        <Pre>
          <code>test</code>
        </Pre>
      );

      const copyButton = screen.getByRole('button', { name: /copy code/i });
      expect(copyButton).toHaveAttribute('title', 'Copy code');
    });

    it('should not be disabled initially', () => {
      render(
        <Pre>
          <code>test</code>
        </Pre>
      );

      const copyButton = screen.getByRole('button', { name: /copy code/i });
      expect(copyButton).not.toBeDisabled();
    });
  });

  describe('Styling', () => {
    it('should have correct CSS classes on button', () => {
      render(
        <Pre>
          <code>test</code>
        </Pre>
      );

      const copyButton = screen.getByRole('button', { name: /copy code/i });

      expect(copyButton.className).toContain('text-foreground-muted');
      expect(copyButton.className).toContain('absolute');
      expect(copyButton.className).toContain('rounded-md');
    });

    it('should have hover styles', () => {
      render(
        <Pre>
          <code>test</code>
        </Pre>
      );

      const copyButton = screen.getByRole('button', { name: /copy code/i });
      expect(copyButton.className).toContain('hover:text-foreground/80');
    });

    it('should have active scale animation', () => {
      render(
        <Pre>
          <code>test</code>
        </Pre>
      );

      const copyButton = screen.getByRole('button', { name: /copy code/i });
      expect(copyButton.className).toContain('active:scale-90');
    });
  });

  describe('Multiple code blocks', () => {
    it('should handle multiple pre blocks independently', async () => {
      render(
        <>
          <Pre>
            <code>first code block</code>
          </Pre>
          <Pre>
            <code>second code block</code>
          </Pre>
        </>
      );

      const buttons = screen.getAllByRole('button', { name: /copy code/i });
      expect(buttons).toHaveLength(2);

      // Click first button
      fireEvent.click(buttons[0]);

      await waitFor(() => {
        expect(mockWriteText).toHaveBeenCalledWith('first code block');
      });
    });
  });

  describe('Edge cases', () => {
    it('should handle empty code block', () => {
      render(
        <Pre>
          <code></code>
        </Pre>
      );

      const copyButton = screen.getByRole('button', { name: /copy code/i });
      expect(copyButton).toBeInTheDocument();
    });

    it('should handle nested elements in code', async () => {
      render(
        <Pre>
          <code>
            <span>const</span> <span>test</span> = <span>true</span>;
          </code>
        </Pre>
      );

      const copyButton = screen.getByRole('button', { name: /copy code/i });
      fireEvent.click(copyButton);

      await waitFor(() => {
        expect(mockWriteText).toHaveBeenCalledWith(expect.stringContaining('const test = true'));
      });
    });
  });

  describe('Snapshot', () => {
    it('should match snapshot in initial state', () => {
      const { container } = render(
        <Pre>
          <code>const test = true;</code>
        </Pre>
      );

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with custom props', () => {
      const { container } = render(
        <Pre className="custom-class" data-language="typescript">
          <code>const test: boolean = true;</code>
        </Pre>
      );

      expect(container).toMatchSnapshot();
    });
  });
});
