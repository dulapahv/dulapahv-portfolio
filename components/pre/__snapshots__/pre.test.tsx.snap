// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pre > Snapshot > should match snapshot in initial state 1`] = `
<div>
  <button
    aria-label="Copy code"
    class="text-foreground-muted absolute top-[5px] right-2 z-50 cursor-pointer rounded-md p-2 !transition-all hover:text-foreground/80 active:scale-90"
    title="Copy code"
  >
    <svg
      class="size-4.5"
      fill="currentColor"
      height="1em"
      viewBox="0 0 256 256"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M216,32H88a8,8,0,0,0-8,8V80H40a8,8,0,0,0-8,8V216a8,8,0,0,0,8,8H168a8,8,0,0,0,8-8V176h40a8,8,0,0,0,8-8V40A8,8,0,0,0,216,32ZM160,208H48V96H160Zm48-48H176V88a8,8,0,0,0-8-8H96V48H208Z"
      />
    </svg>
  </button>
  <pre>
    <code>
      const test = true;
    </code>
  </pre>
</div>
`;

exports[`Pre > Snapshot > should match snapshot with custom props 1`] = `
<div>
  <button
    aria-label="Copy code"
    class="text-foreground-muted absolute top-[5px] right-2 z-50 cursor-pointer rounded-md p-2 !transition-all hover:text-foreground/80 active:scale-90"
    title="Copy code"
  >
    <svg
      class="size-4.5"
      fill="currentColor"
      height="1em"
      viewBox="0 0 256 256"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M216,32H88a8,8,0,0,0-8,8V80H40a8,8,0,0,0-8,8V216a8,8,0,0,0,8,8H168a8,8,0,0,0,8-8V176h40a8,8,0,0,0,8-8V40A8,8,0,0,0,216,32ZM160,208H48V96H160Zm48-48H176V88a8,8,0,0,0-8-8H96V48H208Z"
      />
    </svg>
  </button>
  <pre
    class="custom-class"
    data-language="typescript"
  >
    <code>
      const test: boolean = true;
    </code>
  </pre>
</div>
`;
